{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyO7m8GszmTtfRLLwIKdFWQN",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Sarita0005/bioinformatics-sequence-analysis-python/blob/main/codon_analysis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "_-TETtaInbnb"
      },
      "outputs": [],
      "source": [
        "from google.colab import files\n",
        "uploaded = files.upload()\n",
        "\n",
        "path = '/content/gene.fna'\n",
        "sequence = ''\n",
        "header = ''\n",
        "\n",
        "with open (path, 'r') as f:\n",
        "     for line in f:\n",
        "         if line.startswith('>'):\n",
        "            header =line.strip()\n",
        "         else:\n",
        "             sequence += line.strip()\n",
        "print(\"Header:\")\n",
        "print(header)\n",
        "\n",
        "print(\"\\nSequence length:\")\n",
        "print(len(sequence))"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# gc count percent\n",
        "gc_count = 0\n",
        "total_bases = 0\n",
        "\n",
        "for base in sequence.upper():\n",
        "    if base in ['A','T','G','C']:\n",
        "       total_bases += 1\n",
        "    if base == 'C' or base == 'G':\n",
        "       gc_count += 1\n",
        "\n",
        "gc_percent = (gc_count/ total_bases)*100\n",
        "\n",
        "print('GC Percent:', round(gc_percent,2))"
      ],
      "metadata": {
        "id": "ftyq9nEwnfur"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "a_count =0\n",
        "c_count =0\n",
        "t_count =0\n",
        "g_count =0\n",
        "\n",
        "for base in sequence. upper():\n",
        "    if base == 'A':\n",
        "       a_count += 1\n",
        "    elif base == 'C':\n",
        "        c_count += 1\n",
        "    elif base == 'T':\n",
        "        t_count += 1\n",
        "    elif base == 'G':\n",
        "        g_count += 1\n",
        "total= a_count + g_count + t_count + c_count\n",
        "\n",
        "gc_percent = (g_count + c_count)/total *100\n",
        "at_percent = (a_count + t_count)/total *100\n",
        "gc_at_ratio = (g_count + c_count)/ (a_count + t_count)\n",
        "\n",
        "print('A:', a_count)\n",
        "print('T:', t_count)\n",
        "print('G:', g_count)\n",
        "print('C:', c_count)\n",
        "\n",
        "print('GC Percent:', round(gc_percent,2))\n",
        "\n",
        "print('AT Percent:', round(at_percent,2))\n",
        "print('gc_at_ratio:', round(gc_at_ratio,3))\n"
      ],
      "metadata": {
        "id": "B-Q9KA67nk8Z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "data = {\n",
        "    'Metric':[\n",
        "        'Sequence length',\n",
        "        'A Count',\n",
        "        'T Count',\n",
        "        'C Count',\n",
        "        'G Count',\n",
        "        'GC Percent',\n",
        "        'AT Percent',\n",
        "        'GC/ AT ratio'\n",
        "    ],\n",
        "\n",
        "    'Values':[\n",
        "        total,\n",
        "        a_count,\n",
        "        t_count,\n",
        "        c_count,\n",
        "        g_count,\n",
        "        round(gc_percent,2),\n",
        "        round(at_percent,2),\n",
        "        round(gc_at_ratio,3),\n",
        "\n",
        "    ]\n",
        "}\n",
        "\n",
        "df = pd.DataFrame(data)\n",
        "print(df)\n",
        "\n",
        "# STEP 5: Save results to CSV\n",
        "\n",
        "df.to_csv(\"fasta_sequence_summary.csv\", index=False)\n",
        "\n",
        "print(\"File saved as fasta_sequence_summary.csv\")\n"
      ],
      "metadata": {
        "id": "DxJStEVrnpQ-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Codon count\n",
        "codons =[]\n",
        "for i in range(0,len(sequence)-2,3):\n",
        "     codon = sequence[i:i+3]\n",
        "     codons.append(codon)\n",
        "\n",
        "print('Total Codon:',len(codons))\n",
        "print(codons[:10])"
      ],
      "metadata": {
        "id": "d5UKkkGdnuJ9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from collections import Counter\n",
        "\n",
        "codon_counts = Counter(codons)\n",
        "\n",
        "print(\"Unique codons:\", len(codon_counts))\n",
        "print(codon_counts.most_common(10))"
      ],
      "metadata": {
        "id": "Zy5a5IftnyL6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "codon_to_aa = {\n",
        "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
        "    \"TTA\": \"L\", \"TTG\": \"L\",\n",
        "    \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
        "    \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\",\n",
        "    \"ATG\": \"M\",\n",
        "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
        "\n",
        "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\",\n",
        "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "\n",
        "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
        "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
        "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
        "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
        "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
        "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
        "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
        "\n",
        "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
        "    \"TGG\": \"W\",\n",
        "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
        "    \"AGT\": \"S\", \"AGC\": \"S\",\n",
        "    \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "\n",
        "    \"TAA\": \"*\", \"TAG\": \"*\", \"TGA\": \"*\"\n",
        "}"
      ],
      "metadata": {
        "id": "ImuCLVlGn1iN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "codon_counts_for_aa = codon_counts\n",
        "\n",
        "aa_count = {}\n",
        "\n",
        "for codon, count in codon_counts_for_aa.items():\n",
        "    aa = codon_to_aa.get(codon, \"X\")  # X = unknown\n",
        "    if aa in aa_count:\n",
        "        aa_count[aa] += count\n",
        "    else:\n",
        "        aa_count[aa] = count\n",
        "\n",
        "print(\"Amino Acid Counts:\")\n",
        "for aa, c in aa_count.items():\n",
        "    print(aa, \":\", c)"
      ],
      "metadata": {
        "id": "Ek5NRmd8n9KZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# First, remove stop codons (*) because they are not real amino acids for visualization.\n",
        "aa_count_clean = {aa: c for aa, c in aa_count.items() if aa != \"*\"}\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "plt.figure()\n",
        "plt.bar(aa_count_clean.keys(), aa_count_clean.values())\n",
        "plt.xlabel(\"Amino Acids\")\n",
        "plt.ylabel(\"Count\")\n",
        "plt.title(\"Amino Acid Composition\")\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "TveKn78sn_or"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "total_aa = sum(aa_count_clean.values())\n",
        "\n",
        "aa_freq = {aa: (count / total_aa) * 100\n",
        "           for aa, count in aa_count_clean.items()}\n",
        "plt.figure()\n",
        "plt.bar(aa_freq.keys(), aa_freq.values())\n",
        "plt.xlabel(\"Amino Acids\")\n",
        "plt.ylabel(\"Frequency (%)\")\n",
        "plt.title(\"Amino Acid Frequency (%)\")\n",
        "plt.show()\n",
        "import pandas as pd\n",
        "df_aa = pd.DataFrame({\n",
        "    \"Amino_Acid\": aa_freq.keys(),\n",
        "    \"Frequency (%)\": aa_freq.values()\n",
        "})\n",
        "\n",
        "\n",
        "\n",
        "# sort by frequency\n",
        "df_aa = df_aa.sort_values(by=\"Frequency (%)\", ascending=False)\n",
        "print(df_aa)\n",
        "\n",
        "# save to csv\n",
        "df_aa.to_csv(\"amino_acid_frequency.csv\", index=False)\n",
        "print(\"File saved as amino_acid_frequency.csv\")"
      ],
      "metadata": {
        "id": "-P4ukZ8foKvS"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}